source/extensions/common/wasm/foreign.cc

```diff
@@ -170,8 +170,13 @@ class EvaluateExpressionFactory : public ExpressionFactory {
         ENVOY_LOG(debug, "expr_evaluate error: {}", eval_status.status().message());
         return WasmResult::InternalFailure;
       }
+      auto value = eval_status.value();
+      if (value.IsError()) {
```

Comment:
@TristonianJones is this expected? I clearly saw it in practice.
`eval_status.ok() == true`  but `value.IsError() == true` ?

Reply:
Gads

Reply:
Errors won't necessarily terminate evaluation in CEL. The status_or let's you know whether there is a precondition failure in setting up the expression for evaluation, think a missing function binding. The error indicates whether there was a runtime evaluation error. Since you could combine CEL expression results and error values can be short-circuited away, it's important to preserve the distinction between an application configuration error and an expression evaluation error. We leverage this property extensively in computing an expression across multiple processes

Reply:
How do pre-compilation errors fit in?  Are the execution environment errors all resolved at pre-compilation time or are they reported here as evaluate status (non-value) errors (e.g. because of late binding) ?

Reply:
Stubby has the same issue with application errors, but they never really dealt with it in a consistent way.  This is definitely better than that.

Reply:
You know what, I was not looking very closely at the use of status_or before. This is what I get for trying to grok code on my phone. I was thinking this was the CreateExpression + Evaluate steps combined. A precondition failure normally happens on CreateExpression, whereas Error is an Evaluation concept. 

If Evaluation returns a non-okay status it should signal a CEL framework bug whereas Error is a user-space bug.

Reply:
In this case it was missing activation key, so could be classified as either,  but I understand this is user error now. Thanks.